network_simulator/plot_tool.py

```diff
@@ -0,0 +1,390 @@
+        :param string value_type: the data type of the average.
+        :param boolean count: whether the denominator of the average is a simple
+        count (True), or a timestamp value (False).
+        :param string units: units of the average. e.g. "(sec)"
+        :param list tuple_list: the tuple list to be computed.
+        :return: no return, just print out the statement to stdout.
+        """
+        curr_sum = 0
+        for time, data in tuple_list:
+            curr_sum += data
+        avg = 0
+        if len(tuple_list) != 0:
+            if count:
+                avg = float(curr_sum) / len(tuple_list)
+            else:
+                latest_timestamp = tuple_list[len(tuple_list) - 1][0]
+                avg = float(curr_sum) / latest_timestamp
+        print id + " Average " + value_type + ": " + str(avg) + " " + units
+
+    @staticmethod
+    def output_count_avg_tuple_list(id, value_type, units, tuple_list=[]):
+        """
+        Outputs the average value of a tuple list. This is (count / time).
+
+        Assume that a tuple list is in the format of (timestamp, data), we
+        iterature through the data field of the tuples and outputs the
+        average value to std out.
+
+        :param string id: the Device of this average.
+        :param string value_type: the data type of the average.
+        :param string units: units of the average. e.g. "(sec)"
+        :param list tuple_list: the tuple list to be computed.
+        :return: no return, just print out the statement to stdout.
+        """
+        count = len(tuple_list)
+        avg = 0
+        if count != 0:
+            latest_timestamp = tuple_list[len(tuple_list) - 1][0]
+            avg = float(count) / latest_timestamp
+        print id + " Average " + value_type + ": " + str(avg) + " " + units
+
+
+
+class Grapher(object):
+    """
+    Graphs statistics items from Statistics class.
+    """
+    def __init__(self, stats=None):
+        """
+        :ivar Statistics stats: input Statistics object for graphing.
+        """
+        self.stats = stats
+
+    def graph_links(self):
+        """
+        Graphs the link stats of each available link.
+
+        This includes buffer occupancy, packet losses, and packet transmitted
+        (flow rate).
+        """
+        # Set up legend storage for link names.
+        # Buffer occupancy legend.
+        occpy_legend = []
+        # Packet loss legend.
+        loss_legend = []
+        # Packet transmission legend.
+        trans_legend = []
+
+        # Iterate through each link to display each stats.
+        for link_name, link_stats in self.stats.link_stats.iteritems():
+            # Name of link would be "link_name".
+            # Buffer occupancy w.r.t time.
+            plt.subplot(431)
+            PlotTool().graph_tuple_list(link_stats.buffer_occupancy)
+            occpy_legend.append(link_name)
+            # Output average buffer occupancy to stdout.
```

Comment:
change "stdout" to "log" (or something) in all of the comments that are like this.